test/src/unit-var_offsets.cc

```diff
@@ -0,0 +1,607 @@
+/**
+ * @file   unit-var_offsets.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2020 TileDB Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Tests the different configurations of var-sized attribute offsets
+ */
+
+#include "catch.hpp"
+#include "tiledb/sm/cpp_api/tiledb"
+
+using namespace tiledb;
+
+void create_sparse_array(std::string& array_name) {
```

Comment:
Let's do `const std::string& array_name` here. We typically avoid non-const ref arguments. Same elsewhere in this file where non-const ref arguments are used.

Note that if we truly need a function to mutate an input argument, we prefer raw-pointers:
```
void foo(int* a) {
  *a += 1;
}
```

Reply:
Right, added const for string arguments. However for vector arguments such as in write_dense_array I cannot pass const as set_buffer is expecting non-const.

Reply:
For the vector, we can still use a raw-pointer:
```
void write_sparse_array(
    const std::string& array_name,
    std::vector<int32_t>* data,
    std::vector<uint64_t>* data_offsets) {
  // Write some variable-sized data in the array
  std::vector<int64_t> d1 = {1, 2, 3, 4};
  std::vector<int64_t> d2 = {2, 1, 3, 4};

  Context ctx;
  Array array(ctx, array_name, TILEDB_WRITE);
  Query query(ctx, array, TILEDB_WRITE);
  query.set_layout(TILEDB_UNORDERED)
      .set_buffer("d1", d1)
      .set_buffer("d2", d2)
      .set_buffer("attr", *data_offsets, *data);
  CHECK_NOTHROW(query.submit());

  array.close();
}

...

std::vector<int32_t> data = {1, 2, 3, 4, 5, 6};
std::vector<uint64_t> data_offsets = {0, 4, 12, 20};
write_sparse_array(array_name, &data, &data_offsets);
```

I know this is a little silly since they're functionally equivalent. The motivation is that raw-pointers force the caller to pass the address which forces them to acknowledge that their input variable may be manipulated. It's not a big deal for this test, so you can leave this as-is. Just FYI.